### PR TITLE
PR: Implement Otsu et al. (2018) spectral recovery method

### PR DESCRIPTION
**Pull request by enneract**

### CODE REVIEW

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- colour/recovery/tests/test_otsu2018.py  2
- colour/recovery/datasets/otsu2018.py  8
- colour/recovery/otsu2018.py  3
- colour/recovery/jakob2019.py  21
         

Clones added
============
- colour/recovery/jakob2019.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5836209&bid=19111013)

File: colour/recovery/tests/test_otsu2018.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    definition unit tests methods.
+    """
+
+    def test_roundtrip_colourchecker(self):
+        """
+        Tests :func:`colour.recovery.otsu2018.XYZ_to_sd_otsu2018` definition
+        round-trip errors using a colour checker.
+        """
+
+        for name, sd in COLOURCHECKER_SDS['ColorChecker N Ohta'].items():
```
Comment (KelSolaar) [inline colour/recovery/tests/test_otsu2018.py:44]: Codacy found an issue: [Unused variable 'name'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5836209)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- colour/recovery/datasets/otsu2018.py  8
- colour/recovery/tests/test_otsu2018.py  2
- colour/recovery/jakob2019.py  21
- colour/recovery/otsu2018.py  3
         

Clones added
============
- colour/recovery/jakob2019.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5836209&bid=19111013)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- colour/recovery/otsu2018.py  3
- colour/recovery/datasets/otsu2018.py  8
- colour/recovery/tests/test_otsu2018.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5836209&bid=19111013)

File: colour/recovery/__init__.py [hunk 2]
```diff
@@ -2,6 +2,9 @@
 """
 References
 ----------
+-   :cite:`Otsu2018` : Otsu, H., Yamamoto, M. & Hachisuka, T. (2018)
```
Comment (KelSolaar) [inline colour/recovery/__init__.py:5]: Did I gave you the citation for that? I don't remember but it is perfect! :)

File: colour/recovery/__init__.py [hunk 3]
```diff
@@ -88,7 +94,7 @@ def XYZ_to_sd(XYZ, method='Meng 2015', **kwargs):
 
     References
     ----------
-    :cite:`Meng2015c`, :cite:`Smits1999a`
+    :cite:`Otsu2018`, :cite:`Meng2015c`, :cite:`Smits1999a`
```
Comment (KelSolaar) [inline colour/recovery/__init__.py:97]: In alphabetical order please! :)

File: colour/recovery/datasets/otsu2018.py [hunk 4]
```diff
@@ -0,0 +1,1344 @@
+# -*- coding: utf-8 -*-
+"""
+Otsu et al. (2018) - Reflectance Recovery
+===========================================
```
Comment (KelSolaar) [inline colour/recovery/datasets/otsu2018.py:4]: Pedantry but 2 characters too long I think ;)

File: colour/recovery/otsu2018.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
+# -*- coding: utf-8 -*-
+"""
+Otsu et al. (2018) - Reflectance Recovery
+==============================================
```
Comment (KelSolaar) [inline colour/recovery/otsu2018.py:4]: Bit too long here too :)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- colour/recovery/otsu2018.py  3
- colour/recovery/datasets/otsu2018.py  8
- colour/recovery/tests/test_otsu2018.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5836209&bid=19111013)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- colour/recovery/otsu2018.py  3
- colour/recovery/datasets/otsu2018.py  8
- colour/recovery/tests/test_otsu2018.py  2
         

Complexity decreasing per file
==============================
+ colour/recovery/jakob2019.py  -9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5836209&bid=19111013)

**Comment by enneract** [comment]
I've rebased this branch on top of https://github.com/enneract/colour/commit/c0364967b6346702598017b65689fe9fedf9ccdc. That's develop with the "make jakob2019 work in XYZ" patch applied. I'll be pushing the dataset generation code here very soon.

**Comment by KelSolaar** [comment]
@enneract : Would you mind dropping c036496, please? I cherry-picked it already!

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- colour/recovery/otsu2018.py  12
- colour/recovery/tests/test_otsu2018.py  2
         

Complexity decreasing per file
==============================
+ colour/recovery/jakob2019.py  -9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5836209&bid=19111013)

**Comment by enneract** [comment]
I rebased on top of develop again, dropping https://github.com/colour-science/colour/commit/c0364967b6346702598017b65689fe9fedf9ccdc so it wouldn't show up twice.

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- colour/recovery/otsu2018.py  12
- colour/recovery/tests/test_otsu2018.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5836209&bid=19111013)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- colour/recovery/otsu2018.py  12
- colour/recovery/tests/test_otsu2018.py  4
         

Clones added
============
- colour/recovery/tests/test_otsu2018.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5836209&bid=19111013)

File: colour/recovery/otsu2018.py [hunk 6]
```diff
@@ -0,0 +1,861 @@
...
+        initial_error = self.total_reconstruction_error()
+        _print('Initial error is {:g}.'.format(initial_error))
+
+        for repeat in range(repeats):
+            _print('\n=== Iteration {} of {} ==='.format(repeat + 1, repeats))
+
+            best_total_error = None
+            total_error = self.total_reconstruction_error()
+
+            for i, leaf in enumerate(self.leaves):
```
Comment (KelSolaar) [inline colour/recovery/otsu2018.py:768]: Codacy found an issue: [Unused variable 'i'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5836209)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- colour/recovery/tests/test_otsu2018.py  4
- colour/recovery/otsu2018.py  12
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5836209&bid=19111013)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- colour/recovery/tests/test_otsu2018.py  4
- colour/recovery/otsu2018.py  12
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5836209&bid=19111013)

**Comment by KelSolaar** [comment]
@enneract : Great work on the addendum! Following up, I knew it would take quite a bit to go through it, I took the machete and went clearing up the bushes a bit and ensured the code was as much as possible consistent with the codebase: https://github.com/colour-science/colour/commit/cbaf952e2999f1bc6ffca492e48dbe35237a181f

The branch is rebased against the very latest develop.

I would suggest not looking at the diff which is large but the final code. I did not document everything but the gist of it is as follows:

## Overall

- Removed `Otsu2018Error`
- Add various `__str__` and `__len__` methods.
- Added some regression tests for the class

## Otsu2018Dataset

- Renamed `Otsu2018Dataset` to `Dataset_Otsu2018`
- Add properties to `Dataset_Otsu2018`
- Renamed `Otsu2018Dataset.to_file` to `Dataset_Otsu2018.write`
- Renamed `Otsu2018Dataset.from_file` to `Dataset_Otsu2018.read`
- Removed `Otsu2018Dataset.to_Python_file`

## DATASET_REFERENCE_OTSU2018

- Renamed `builtin_Otsu2018_dataset` to `DATASET_REFERENCE_OTSU2018`

## XYZ_to_sd_Otsu2018

- Ensured that `cmfs` is used in `XYZ_to_sd_Otsu2018`

## PartitionAxis

- Converted `PartitionAxis` to a `namedtuple` class

## ColourData

- Renamed `Colours` to `ColourData`
- `ColourData.__init__` uses `msds_to_XYZ`, from 3 sec to 0.01 sec
- Add properties to `ColourData`

## Node

- Renamed `Node.leaf` to `Node.is_leaf`
- Removed `Node._reconstruction_xy` not worth introducing another method to avoid an `xy` conversion
- Optimised `Node.PCA`
- Removed `Node._PCA_done` only really need to check `self._M_inverse`
- Renamed `Node.reconstruction_error` to `Node.leaf_reconstruction_error`
- Renamed `Node.total_reconstruction_error` to `Node.branch_reconstruction_error`
- Renamed `Node.partition_error` to `Node.partition_reconstruction_error`
- Update progress bar management and added support for **tqdm.**

## NodeTree_Otsu2018

- Renamed `Otsu2018Tree` to `NodeTree_Otsu2018`
- Add properties to `NodeTree_Otsu2018`
- Removed `self._progress_bar`
- Removed `self.xy_w` as it was not used
- Removed `_print` in `NodeTree_Otsu2018` optimise as one can pass directly a `logging` function here with stamping and anything one can dream of

---

If you are happy with that, feel free to push into your branch, add remaining bits of polish and we can merge!

File: colour/utilities/common.py [hunk 7]
```diff
@@ -442,6 +442,41 @@ def is_pandas_installed(raise_exception=False):
...
+        return True
+    except ImportError as error:  # pragma: no cover
+        if raise_exception:
+            raise ImportError(
+                ('"tqdm" related API features are not available: "{0}".\n'
+                 'Please refer to the installation guide for more '
+                 'information: '
+                 'https://www.colour-science.org/installation-guide/'
+                 ).format(error))
+            return False
```
Comment (KelSolaar) [inline colour/utilities/common.py:477]: Codacy found an issue: [Unreachable code](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5836209)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- colour/recovery/otsu2018.py  10
- colour/recovery/tests/test_otsu2018.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5836209&bid=19111013)

File: colour/utilities/common.py [hunk 8]
```diff
@@ -442,6 +442,41 @@ def is_pandas_installed(raise_exception=False):
...
+        Is *tqdm* installed.
+
+    Raises
+    ------
+    ImportError
+        If *tqdm* is not installed.
+    """
+
+    try:  # pragma: no cover
+        import tqdm  # noqa
```
Comment (KelSolaar) [inline colour/utilities/common.py:466]: Codacy found an issue: [Unused variable 'tqdm'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5836209)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- colour/recovery/otsu2018.py  10
- colour/recovery/tests/test_otsu2018.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5836209&bid=19111013)

File: colour/utilities/common.py [hunk 9]
```diff
@@ -442,17 +442,53 @@ def is_pandas_installed(raise_exception=False):
...
```
Comment (KelSolaar) [inline colour/utilities/common.py:466]: Codacy found an issue: [Unused variable 'tqdm'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5836209)

Comment (KelSolaar) [reply colour/utilities/common.py:466]: Will fix post-merge!

**Comment by KelSolaar** [comment]
Thanks a lot @enneract! 👍

Merging :)

**Pull request merged**